controllers/apply.go

```diff
@@ -69,19 +101,170 @@ func getK8sSensorDeployments(
 	statusManager status.AgentStatusManager,
 	k8SensorBackends []backends.K8SensorBackend,
 	keysSecret *corev1.Secret,
+	deploymentContext *k8ssensordeployment.DeploymentContext,
 ) []builder.ObjectBuilder {
 	builders := make([]builder.ObjectBuilder, 0, len(k8SensorBackends))
 
 	for _, backend := range k8SensorBackends {
 		builders = append(
 			builders,
-			k8ssensordeployment.NewDeploymentBuilder(agent, isOpenShift, statusManager, backend, keysSecret),
+			k8ssensordeployment.NewDeploymentBuilder(agent, isOpenShift, statusManager, backend, keysSecret, deploymentContext),
 		)
 	}
 
 	return builders
 }
 
+// getSortedTargets returns a sorted copy of the given targets slice
+func getSortedTargets(targets []string) []string {
+	sortedTargets := make([]string, len(targets))
+	copy(sortedTargets, targets)
+	sort.Strings(sortedTargets)
+	return sortedTargets
+}
+
+// compareAndUpdateETCDTargets compares current ETCD targets with new discovered targets
+// and determines if an update is needed. This function handles environment variable extraction,
+// target sorting, and comparison logic.
+func compareAndUpdateETCDTargets(
+	existingDeployment *appsv1.Deployment,
+	discoveredTargets []string,
+	log logr.Logger,
+) bool {
+	log.Info(
+		"Comparing current ETCD targets with discovered targets to determine if update is needed",
+	)
+
+	// Extract current ETCD targets from deployment environment variables
+	currentTargets := ""
+	for _, container := range existingDeployment.Spec.Template.Spec.Containers {
+		if container.Name == constants.ContainerK8Sensor {
+			for _, env := range container.Env {
+				if env.Name == constants.EnvETCDTargets {
+					currentTargets = env.Value
+					break
+				}
+			}
+			break
+		}
+	}
+
+	// Sort discovered targets to ensure consistent comparison
+	sortedDiscoveredTargets := getSortedTargets(discoveredTargets)
+	newTargets := strings.Join(sortedDiscoveredTargets, ",")
+
+	// Sort current targets for proper comparison
+	if currentTargets != "" {
+		currentTargetsList := strings.Split(currentTargets, ",")
+		sort.Strings(currentTargetsList)
+		currentTargets = strings.Join(currentTargetsList, ",")
+	}
+
+	log.Info("Target comparison details",
+		"currentTargets", currentTargets,
+		"newTargets", newTargets,
+		"needsUpdate", currentTargets != newTargets)
+
+	// Return true if targets are different (update needed)
+	return currentTargets != newTargets
+}
+
+// CreateDeploymentContext creates a deployment context for the k8s-sensor deployment.
+// It handles both OpenShift and vanilla Kubernetes cases, setting up the appropriate
+// ETCD configuration based on the environment.
+func CreateDeploymentContext(
+	ctx context.Context,
+	deps DeploymentContextDependencies,
+	agent *instanav1.InstanaAgent,
```

Comment:
The idea was that you would pass the client not an agent or this deps struct/interface.